From c85fa64b36f21540c27d8c6fa53e8f719d2f3c8d Mon Sep 17 00:00:00 2001
From: Zhang Qiao <zhangqiao22@huawei.com>
Date: Wed, 8 Jun 2022 17:30:12 +0000
Subject: [PATCH 3324/4585] [Huawei] sched/fair: Fix one cfs task run on a cpu
 with many IRQS

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 186981

--------------------------------

In some cases, the busiest group's utilization is max or even
higher than capacity because of IRQs, so if local cpu is idle,
and local group capacity is greater busiest group, we try pull
a task.

This optimization is controlled by sched_features LB_BUSY.

Signed-off-by: Zhang Qiao <zhangqiao22@huawei.com>
Signed-off-by: Zheng Zucheng <zhengzucheng@huawei.com>
---
 kernel/sched/fair.c     | 14 ++++++++++++--
 kernel/sched/features.h |  1 +
 2 files changed, 13 insertions(+), 2 deletions(-)

diff --git a/kernel/sched/fair.c b/kernel/sched/fair.c
index f651ae268d9b..0ae592f9ffc2 100644
--- a/kernel/sched/fair.c
+++ b/kernel/sched/fair.c
@@ -10073,6 +10073,13 @@ static inline void calculate_imbalance(struct lb_env *env, struct sd_lb_stats *s
 			env->migration_type = migrate_task;
 			lsub_positive(&nr_diff, local->sum_nr_running);
 			env->imbalance = nr_diff >> 1;
+			if (sched_feat(LB_BUSY)) {
+				if (env->idle != CPU_NOT_IDLE && env->imbalance == 0 &&
+					busiest->group_type == group_fully_busy &&
+					(busiest->group_capacity * env->sd->imbalance_pct <=
+					 local->group_capacity * 100))
+					env->imbalance = 1;
+			}
 		} else {
 
 			/*
@@ -10275,11 +10282,14 @@ static struct sched_group *find_busiest_group(struct lb_env *env)
 			 */
 			goto out_balanced;
 
-		if (busiest->sum_h_nr_running == 1)
+		if (busiest->sum_h_nr_running == 1) {
 			/*
 			 * busiest doesn't have any tasks waiting to run
 			 */
-			goto out_balanced;
+			if (!sched_feat(LB_BUSY) ||
+				busiest->group_type != group_fully_busy)
+				goto out_balanced;
+		}
 	}
 
 force_balance:
diff --git a/kernel/sched/features.h b/kernel/sched/features.h
index b358bbcb800b..c7f063ea136b 100644
--- a/kernel/sched/features.h
+++ b/kernel/sched/features.h
@@ -107,3 +107,4 @@ SCHED_FEAT(BASE_SLICE, true)
  */
 SCHED_FEAT(DA_UTIL_TASKGROUP, true)
 #endif
+SCHED_FEAT(LB_BUSY, false)
-- 
2.17.1

